File: ps/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* zero = new ListNode(0);
        ListNode* res = new ListNode(-1);
        ListNode* cur = res;
        bool carry = false;
        while (l1 || l2 || carry) {
            if (!l1) { l1 = zero;}
            if (!l2) { l2 = zero;}
            cur->next = new ListNode(l1->val + l2->val);
            cur = cur->next;
            cur->val += carry ? 1 : 0;
            carry = (cur->val >= 10) ? true : false;
            cur->val %= 10;
            l1 = l1->next;
            l2 = l2->next;
        }
        return res->next;
    }
};

File: ps/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        dummy_head = ListNode()
        node = dummy_head
        while l1 and l2:

            new_node = ListNode((l1.val + l2.val + carry) % 10)
            if l1.val + l2.val + carry >= 10:
                carry = 1
            else:
                carry = 0
            node.next = new_node
            node = node.next
            l1 = l1.next
            l2 = l2.next

        if not l1:
            l1 = l2
        while l1:
            new_node = ListNode((l1.val + carry) % 10)
            if l1.val + carry >= 10:
                carry = 1
            else:
                carry = 0
            node.next = new_node
            node = node.next
            l1 = l1.next

        if carry == 1:
            node.next = ListNode(1)

        return dummy_head.next